File: 271-encode-and-decode-strings/271-encode-and-decode-strings.py
class Codec:

    def encode(self, strs):
        if strs == [""]:
            return ""
        encoded = []
        for word in strs:
            encoded.append(str(len(word)))
            encoded.append("#")
            encoded.append(word)
        return "".join(encoded)
        

    def decode(self, s):
        if s == "":
            return [""]
        decoded = []
        i = 0
        while i < len(s):
            startLength = i 
            while s[i] != "#":
                i += 1
            endLength = i
            length = int(s[startLength:endLength]) 
            start = i + 1 
            end = length + start 
            decoded.append(s[start:end]) 
            i = end
        return decoded
                
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))

#["63/Rc","h","BmI3FS~J9#vmk","7uBZ?7*/","24h+X","O "]
#"5#63/Rc1h13#BmI3FS~J9#vmk"

File: 271-encode-and-decode-strings/271-encode-and-decode-strings.js
/**
 * Encodes a list of strings to a single string.
 *
 * @param {string[]} strs
 * @return {string}
 */
var encode = function(strs) {
    if (strs === "") return ""
    const res = []
    for (const word of strs) {
        res.push(word.length.toString(), "#")
        for (const letter of word) {
            res.push(letter)
        } 
    }
    return res.join("")
};

/**
 * Decodes a single string to a list of strings.
 *
 * @param {string} s
 * @return {string[]}
 */
var decode = function(s) {
    if (s === "") return [""]
    const res = []
    let i = 0
    while (i < s.length) {
        const j = i
        while (s[i] !== "#") {
            i++
        }
        const count = parseInt(s.slice(j, i))
        i += 1
        res.push(s.slice(i, i + count))
        i += count 
    }
    return res
};

/**
 * Your functions will be called as such:
 * decode(encode(strs));
 */